File: LeetCode/rotate_list.rb
class Node < Struct.new(:value, :link)
end

class List
    def initialize
        @head = nil
        @size = 0;
    end

    def add(value)
        @head = Node.new(value, @head)
        @size += 1
    end

    def rotate(k)
        if k > @size
            return
        end

        it = @head
        (@size - k - 1).times {
            it = it.link
        }

        last = it
        while last.link
            last = last.link
        end

        old_head = @head
        @head = it.link
        it.link = nil

        last.link = old_head
    end

    def inspect
        s = ""
        it = @head
        while it
            s << "#{it.value} "
            it = it.link
        end
        return s
    end
end

list = List.new
5.downto(1).each { |x|
    list.add(x)
}

puts list.inspect
list.rotate(2)
puts list.inspect

File: LeetCode/rotate_list.cs
using System;

class Program {
    class Node {
        public int Data { get; set; }
        public Node Next { get; set; }
    
        public override String ToString() {
            return String.Format("{0} {1}", Data, Next == null ? String.Empty : Next.ToString());
        }

        public Node Rotate(int k) {
            Node last = this;
            Node kLast = this;
            for (int i = 0; i < k; i++) {
                kLast = kLast.Next;
            }

            while (kLast.Next != null) {
               kLast = kLast.Next;
               last = last.Next; 
            }

            Node result = last.Next;
            last.Next = null;
            kLast.Next = this;
            return result;
        }
    }

    static void Main() {
        Node head = null;
        head = new Node { Data = 5, Next = head };
        head = new Node { Data = 4, Next = head };
        head = new Node { Data = 3, Next = head };
        head = new Node { Data = 2, Next = head };
        head = new Node { Data = 1, Next = head };
        Console.WriteLine(head);
        Console.WriteLine(head.Rotate(2));
    }
}


